### ISSUE TITLE
Authentication with Session / User instead of api token

### ISSUE THREAD

**Issue by Graphicscore** [pimcore/data-hub#205]
Hi,
I was wondering if there is any support for authenticating GraphQL / Webservice queries with a pimcore user / session. I don't want to use a static api key but would like to use some sort of session tokens that will be acquired through a login.

Did I miss something or is there a trick ? 😄

**Comment by julmot** [comment]
Related to: #156

**Comment by solverat** [comment]
Using sessions can be a real performance killer. Using something like JWT should be the better approach while working with api frameworks right?

**Comment by Graphicscore** [comment]
@solverat I would be fine with using JWT. I would like to restrict access to certain parts of the API and/or  return data based on who is doing the request. For example having multiple users with different usergroups and return different data depending on which usergroup they have. Is there something like this planned or is there some documentation on how I would do that ?

**Comment by julmot** [comment]
@Graphicscore I love that someone brings this up. I've already asked for this on the last Pimcore conference.

I agree, would be great to consider personalized content in this regard too. DataHub should step by step offer all of the Pimcore specific features (also e.g. Tag & Snippet Management, etc.) in the GraphQl API too.

**Comment by rajesh17692** [comment]
For this, you can override GraphQL WebserviceController and in that, you can add extra security check-in performSecurityCheck function like below

$apiKey = $request->headers->get('X-Api-Key');
            $token = $request->headers->get('token');
            $userObj = User::getByUserToken($token,1);
            if ($apiKey === $securityConfig['apikey']) {
                if ($userObj) {
                    if($userObj->getUserToken() === $token){
                        return;
                    }else {
                        throw new AccessDeniedHttpException('invalid token');
                    }
                } else {
                    throw new AccessDeniedHttpException('invalid token');
                }
            }else{
                $apiKey = $request->get('apikey');
                if ($apiKey === $securityConfig['apikey']) {
                    return;
                }
            }

**Comment by pimcoremohit31** [comment]
@rajesh17692 
have you integrated JWT with Graph-Ql in Pimcore?

**Comment by rajesh17692** [comment]
Yes

**Comment by pimcoremohit31** [comment]
@rajesh17692 
That's great...
I am also using JWT with Graph-Ql and I want to authenticate with JWT. Can you help me if you can provide me the configuration code?
Actually, I have used your code in my code but I am not able to call this code `$userObj = User::getByUserToken($token,1);`.
I mean how I will get the token from User class? 
Can you please explain in detail?
Thank you in advance.

**Comment by rajesh17692** [comment]
@pimcoremohit31  I am storing token of user with data when the user trying to log in so with the next request we can compare. token.

**Comment by pimcoremohit31** [comment]
@rajesh17692 Okay, Then how you are handling the expiring time for the token?

**Comment by rajesh17692** [comment]
@pimcoremohit31  sorry I am using below token generation method

Csrf\TokenGenerator\UriSafeTokenGenerator();

**Comment by pimcoremohit31** [comment]
@rajesh17692  Okay. So, you are not using the JWT bundle?

**Comment by rajesh17692** [comment]
No. You can user JWT instead and to validate the token with every request you can create helper function and validate after geting data from user account

**Comment by pimcoremohit31** [comment]
Okay..Thank you @rajesh17692  for you help.
